Packages/com.unity.inputsystem/InputSystem/Devices/InputDevice.cs

```diff
@@ -446,10 +446,53 @@ internal bool hasControlsWithDefaultState
         internal void SetUsage(InternedString usage)
         {
             // Make last entry in m_UsagesForEachControl be our device usage string.
+            SetUsage(m_UsageToControl != null ? m_UsageToControl.Length : 0, usage);
+        }
+
+        internal void AddUsage(InternedString usage)
+        {
+            // Check if the usage already exists
+            if (Array.Exists(m_UsagesForEachControl, o => o == usage))
+                return;
+
+            SetUsage(m_UsagesForEachControl != null ? m_UsagesForEachControl.Length : 0, usage);
+        }
+
+        internal void RemoveUsage(InternedString usage)
+        {
             var numControlUsages = m_UsageToControl != null ? m_UsageToControl.Length : 0;
-            Array.Resize(ref m_UsagesForEachControl, numControlUsages + 1);
-            m_UsagesForEachControl[numControlUsages] = usage;
-            m_UsagesReadOnly = new ReadOnlyArray<InternedString>(m_UsagesForEachControl, numControlUsages, 1);
+            var usageIndex = Array.IndexOf(m_UsagesForEachControl, usage, numControlUsages);
```

Comment:
Would prefer not to even iterate over the usages of the controls and only iterate over the ones for the device instead.

```
var index = -1;
var usages = this.usages;
for (var i = 0; i < usages.Count; ++i)
    if (usages[i] == usage)
    {
        index = i;
        break;
    }
if (index == -1)
    return;
```